### PR TITLE
Throw away sasl client/server handlers after authentication

### PR DESCRIPTION
**Pull request by ggezer**

### CODE REVIEW

**Comment by ggezer** [comment]
@bf8086 PTAL. Thanks.

File: core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java [hunk 1]
```diff
@@ -81,9 +82,15 @@ public void onNext(SaslMessage saslMessage) {
...
       // Because client will assume authenticated after receiving the success message.
       if (response.getMessageType() == SaslMessageType.SUCCESS) {
+        // Register authorized user with the authentication server.
         mAuthenticationServer.registerChannel(mChannelId,
-            mSaslServerHandler.getAuthenticatedUserInfo(), this,
-            mSaslServerHandler.getSaslServer());
+            mSaslServerHandler.getAuthenticatedUserInfo(), this);
+        // Finished with the handler.
+        try {
+          mSaslServerHandler.close();
```
Comment (bf8086) [inline core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java]: Should we also close the handle when channel is not authenticated successfully?

Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java]: Definitely. Fixed.

File: core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java [hunk 2]
```diff
@@ -111,6 +120,13 @@ public void onError(Throwable throwable) {
       LOG.warn("Closing authenticated channel: {} due to error: {}", mChannelId, throwable);
       mAuthenticationServer.unregisterChannel(mChannelId);
     }
+    if (mSaslServerHandler != null) {
+      try {
+        mSaslServerHandler.close();
+      } catch (IOException exc) {
```
Comment (bf8086) [inline core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java]: Do we also need to close the handler when `SaslStreamServerDriver.close()` is called?

**Comment by ggezer** [comment]
Jenkins, test this please.

File: core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java (hunk 2 above)
Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java]: No, SaslServer will be closed before reaching there for both success and failure paths.

Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/SaslStreamServerDriver.java]: I couldn't find any cleanar way for serverdriver to close its handler. Now closing it during close call as well.

File: core/common/src/main/java/alluxio/security/authentication/plain/SaslClientHandlerPlain.java [hunk 3]
```diff
@@ -92,4 +96,15 @@ public ChannelAuthenticationScheme getClientScheme() {
...
     return mSaslClient;
   }
+
+  @Override
+  public void close() {
+    if (mSaslClient != null) {
+      try {
+        mSaslClient.dispose();
+      } catch (SaslException exc) {
+        LOG.debug("Failed to close SaslClient.", exc);
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/security/authentication/plain/SaslClientHandlerPlain.java]: Is it a problem if we fail to close the client or server?

Comment (ggezer) [reply core/common/src/main/java/alluxio/security/authentication/plain/SaslClientHandlerPlain.java]: Not really. But I don't expect that to happen often, that's why the debug trace.

**Comment by calvinjia** [comment]
alluxio-bot, merge this please.

**Comment by calvinjia** [comment]
alluxio-bot, cherry-pick this to branch-2.0 please.

**Pull request merged**